src/main/java/betteradvancements/util/CriterionGrid.java

```diff
@@ -0,0 +1,132 @@
+package betteradvancements.util;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import net.minecraft.advancements.Advancement;
+import net.minecraft.advancements.AdvancementProgress;
+import net.minecraft.advancements.Criterion;
+import net.minecraft.client.gui.FontRenderer;
+
+public class CriterionGrid {
+    public static boolean showSpoilers = false;
+
+    private final List<String> cellContents;
+    private final int[] cellWidths;
+    private final int fontHeight;
+    private final int numColumns;
+    public final int numRows;
+    public List<Column> columns;
+    public int width;
+    public int height;
+    public double aspectRatio;
+
+    public CriterionGrid() {
+        this.cellContents = Collections.emptyList();
+        this.cellWidths = new int[0];
+        this.fontHeight = 0;
+        this.numColumns = 0;
+        this.numRows = 0;
+        columns = Collections.emptyList();
+        width = 0;
+        height = 0;
+        aspectRatio = Double.NaN;
+    }
+
+    public CriterionGrid(List<String> cellContents, int[] cellWidths, int fontHeight, int numColumns) {
+        this.cellContents = cellContents;
+        this.cellWidths = cellWidths;
+        this.fontHeight = fontHeight;
+        this.numColumns = numColumns;
+        this.numRows = (int)Math.ceil((double)cellContents.size() / numColumns);
+    }
+
+    public void init() {
+        List<Column> columns = new ArrayList<Column>();

```

Comment:
Does this really need to be a new variable? Couldn't you work in `this.columns`.

Reply:
Usually I like to avoid populating non-local variables with partially complete data, but in this case we're not multi-threading so there should be no problem making this change.